### DISCUSSION TITLE
How to enable trace/logging for lemminx extension run within WWD XML LS process?

### DISCUSSION THREAD

**Discussion by scottkurz** [mickaelistria/eclipse-bluesky#974] [General]
Hi,  I'm trying to enable trace/logging for a lemminx extension running in XML LS launched from Eclipse 2022-09 using the wildwebdeveloper XML function.   The lemminx XML LS is launched as a separate JVM process in my env.    The extension is now coded with java.util.logging.     

I see there is a "hook" to add a debug port [here](https://github.com/eclipse/wildwebdeveloper/blob/d913c31661c5bff06615baa7799d504e037c22b9/org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/XMLLanguageServer.java#L92) but I don't see something comparable for say a JUL properties file property.

Also I know we can use the preference of **Language Servers -> Logs** to trace the JSON-RPC calls.. but I'm trying to trace the "internals" not just the JSON-RPC payloads. 

Would this be new function needed?  Or is there another way to accomplish this?  Thank you!

**Comment by mickaelistria** [comment]
Indeed, such an contribution allowing to pass some specific debugging configuration to LemMinX would be welcome.

**Comment by scottkurz** [comment]
It looks like the lemminx side has an opening to [parameterize ](https://github.com/eclipse/lemminx/blob/b11ed6b6fa20e141ef7bbe28b883b5b51986be13/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/XMLLanguageServer.java#L114) it with logging-related settings.  

At first it superficially seems like the WWD XML extension point's `<initializationOptionsProvider>` [element](https://github.com/eclipse/wildwebdeveloper/blob/d913c31661c5bff06615baa7799d504e037c22b9/org.eclipse.wildwebdeveloper.xml/schema/lemminxExtension.exsd#L84) might have been intended to align with this lemminx feature (because of the names sounding similar).

Browsing the code, it's not clear to me that the two are really connected in this way, however.  Since it would take some trial and error to set up all the XML/Java/JSON artifacts and content here, I'm going to stop there and see what anyone thinks about this.

---

Referring back to the lemminx XMLLanguageServer class, I would've hoped we could at least have leveraged the default JRE JUL config.  However, the lemminx server proceeds to [unregister](https://github.com/eclipse/lemminx/blob/b11ed6b6fa20e141ef7bbe28b883b5b51986be13/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/logs/LogHelper.java#L48) all handlers.   If it weren't for this I'd have been happy for now to have the INFO and higher messages logged to ConsoleHandler, and have this output stream redirected back to the parent Eclipse IDE (using `-consoleLog`).

In the meantime, I'm realizing I can workaround this by adding `ConsoleHandler` back to the root Logger within my extension programmatically (if I don't mind overwriting whatever's been intended to be configured for the rest of Lemminx plus extensions).

**Reply by mickaelistria** [reply]
See also https://github.com/eclipse/lemminx-maven/issues/300 .
It seems like LemMinX adds a layer of indirection/complexity here. I was hoping just adding a `-Djava.util.logging.config.file=...` in the command generated by `org.eclipse.wildwebdeveloper.xml.internal.XMLLanguageServer` would be sufficient.

**Comment by mickaelistria** [comment]
https://github.com/eclipse/wildwebdeveloper/pull/1040 aims at allowing LemMinX server logs.

**Reply by scottkurz** [reply]
OK, I added a note to the doc https://github.com/eclipse/wildwebdeveloper/pull/1094 to explain this.

**Comment by scottkurz** [comment]
Added a PR to lemminx to use the `log.level` property value as the level to set on the lemminx-created FileHandler :  https://github.com/eclipse/lemminx/pull/1482
